File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-ird-5-4/CapFloor.cpp
// CapFloor.cpp 
#include "CapFloor.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CapFloor::CapFloor(TiXmlNode* xmlNode)
: Product(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //capFloorStreamNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* capFloorStreamNode = xmlNode->FirstChildElement("capFloorStream");

   if(capFloorStreamNode){capFloorStreamIsNull_ = false;}
   else{capFloorStreamIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- capFloorStreamNode , address : " << capFloorStreamNode << std::endl;
   #endif
   if(capFloorStreamNode)
   {
      if(capFloorStreamNode->Attribute("href") || capFloorStreamNode->Attribute("id"))
      {
          if(capFloorStreamNode->Attribute("id"))
          {
             capFloorStreamIDRef_ = capFloorStreamNode->Attribute("id");
             capFloorStream_ = boost::shared_ptr<InterestRateStream>(new InterestRateStream(capFloorStreamNode));
             capFloorStream_->setID(capFloorStreamIDRef_);
             IDManager::instance().setID(capFloorStreamIDRef_,capFloorStream_);
          }
          else if(capFloorStreamNode->Attribute("href")) { capFloorStreamIDRef_ = capFloorStreamNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { capFloorStream_ = boost::shared_ptr<InterestRateStream>(new InterestRateStream(capFloorStreamNode));}
   }

   //premiumNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* premiumNode = xmlNode->FirstChildElement("premium");

   if(premiumNode){premiumIsNull_ = false;}
   else{premiumIsNull_ = true;}

   if(premiumNode)
   {
      for(premiumNode; premiumNode; premiumNode = premiumNode->NextSiblingElement("premium")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- premiumNode , address : " << premiumNode << std::endl;
          #endif
          if(premiumNode->Attribute("href") || premiumNode->Attribute("id"))
          {
              if(premiumNode->Attribute("id"))
              {
                  premiumIDRef_ = premiumNode->Attribute("id");
                  premium_.push_back(boost::shared_ptr<Payment>(new Payment(premiumNode)));
                  premium_.back()->setID(premiumIDRef_);
                  IDManager::instance().setID(premiumIDRef_, premium_.back());
              }
              else if(premiumNode->Attribute("href")) { premiumIDRef_ = premiumNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { premium_.push_back(boost::shared_ptr<Payment>(new Payment(premiumNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- premiumNode , address : " << premiumNode << std::endl;
       #endif
   }

   //additionalPaymentNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* additionalPaymentNode = xmlNode->FirstChildElement("additionalPayment");

   if(additionalPaymentNode){additionalPaymentIsNull_ = false;}
   else{additionalPaymentIsNull_ = true;}

   if(additionalPaymentNode)
   {
      for(additionalPaymentNode; additionalPaymentNode; additionalPaymentNode = additionalPaymentNode->NextSiblingElement("additionalPayment")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- additionalPaymentNode , address : " << additionalPaymentNode << std::endl;
          #endif
          if(additionalPaymentNode->Attribute("href") || additionalPaymentNode->Attribute("id"))
          {
              if(additionalPaymentNode->Attribute("id"))
              {
                  additionalPaymentIDRef_ = additionalPaymentNode->Attribute("id");
                  additionalPayment_.push_back(boost::shared_ptr<Payment>(new Payment(additionalPaymentNode)));
                  additionalPayment_.back()->setID(additionalPaymentIDRef_);
                  IDManager::instance().setID(additionalPaymentIDRef_, additionalPayment_.back());
              }
              else if(additionalPaymentNode->Attribute("href")) { additionalPaymentIDRef_ = additionalPaymentNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { additionalPayment_.push_back(boost::shared_ptr<Payment>(new Payment(additionalPaymentNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- additionalPaymentNode , address : " << additionalPaymentNode << std::endl;
       #endif
   }

   //earlyTerminationProvisionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* earlyTerminationProvisionNode = xmlNode->FirstChildElement("earlyTerminationProvision");

   if(earlyTerminationProvisionNode){earlyTerminationProvisionIsNull_ = false;}
   else{earlyTerminationProvisionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- earlyTerminationProvisionNode , address : " << earlyTerminationProvisionNode << std::endl;
   #endif
   if(earlyTerminationProvisionNode)
   {
      if(earlyTerminationProvisionNode->Attribute("href") || earlyTerminationProvisionNode->Attribute("id"))
      {
          if(earlyTerminationProvisionNode->Attribute("id"))
          {
             earlyTerminationProvisionIDRef_ = earlyTerminationProvisionNode->Attribute("id");
             earlyTerminationProvision_ = boost::shared_ptr<EarlyTerminationProvision>(new EarlyTerminationProvision(earlyTerminationProvisionNode));
             earlyTerminationProvision_->setID(earlyTerminationProvisionIDRef_);
             IDManager::instance().setID(earlyTerminationProvisionIDRef_,earlyTerminationProvision_);
          }
          else if(earlyTerminationProvisionNode->Attribute("href")) { earlyTerminationProvisionIDRef_ = earlyTerminationProvisionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { earlyTerminationProvision_ = boost::shared_ptr<EarlyTerminationProvision>(new EarlyTerminationProvision(earlyTerminationProvisionNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<InterestRateStream> CapFloor::getCapFloorStream()
{
   if(!this->capFloorStreamIsNull_){
        if(capFloorStreamIDRef_ != ""){
             return boost::shared_static_cast<InterestRateStream>(IDManager::instance().getID(capFloorStreamIDRef_));
        }else{
             return this->capFloorStream_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<InterestRateStream>();
   }
}
std::vector<boost::shared_ptr<Payment>> CapFloor::getPremium()
{
   if(!this->premiumIsNull_){
        if(premiumIDRef_ != ""){
             return this->premium_;
        }else{
             return this->premium_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<Payment>>();
   }
}
std::vector<boost::shared_ptr<Payment>> CapFloor::getAdditionalPayment()
{
   if(!this->additionalPaymentIsNull_){
        if(additionalPaymentIDRef_ != ""){
             return this->additionalPayment_;
        }else{
             return this->additionalPayment_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<Payment>>();
   }
}
boost::shared_ptr<EarlyTerminationProvision> CapFloor::getEarlyTerminationProvision()
{
   if(!this->earlyTerminationProvisionIsNull_){
        if(earlyTerminationProvisionIDRef_ != ""){
             return boost::shared_static_cast<EarlyTerminationProvision>(IDManager::instance().getID(earlyTerminationProvisionIDRef_));
        }else{
             return this->earlyTerminationProvision_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EarlyTerminationProvision>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-ird-5-4/CapFloor.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CapFloor : Product
    {
        public CapFloor(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode capFloorStreamNode = xmlNode.SelectSingleNode("capFloorStream");
            
            if (capFloorStreamNode != null)
            {
                if (capFloorStreamNode.Attributes["href"] != null || capFloorStreamNode.Attributes["id"] != null) 
                {
                    if (capFloorStreamNode.Attributes["id"] != null) 
                    {
                        capFloorStreamIDRef_ = capFloorStreamNode.Attributes["id"].Value;
                        InterestRateStream ob = new InterestRateStream(capFloorStreamNode);
                        IDManager.SetID(capFloorStreamIDRef_, ob);
                    }
                    else if (capFloorStreamNode.Attributes["href"] != null)
                    {
                        capFloorStreamIDRef_ = capFloorStreamNode.Attributes["href"].Value;
                    }
                    else
                    {
                        capFloorStream_ = new InterestRateStream(capFloorStreamNode);
                    }
                }
                else
                {
                    capFloorStream_ = new InterestRateStream(capFloorStreamNode);
                }
            }
            
        
            XmlNodeList premiumNodeList = xmlNode.SelectNodes("premium");
            
            if (premiumNodeList != null)
            {
                this.premium_ = new List<Payment>();
                foreach (XmlNode item in premiumNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            premiumIDRef_ = item.Attributes["id"].Value;
                            premium_.Add(new Payment(item));
                            IDManager.SetID(premiumIDRef_, premium_[premium_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            premiumIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        premium_.Add(new Payment(item));
                        }
                    }
                    else
                    {
                        premium_.Add(new Payment(item));
                    }
                }
            }
            
        
            XmlNodeList additionalPaymentNodeList = xmlNode.SelectNodes("additionalPayment");
            
            if (additionalPaymentNodeList != null)
            {
                this.additionalPayment_ = new List<Payment>();
                foreach (XmlNode item in additionalPaymentNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            additionalPaymentIDRef_ = item.Attributes["id"].Value;
                            additionalPayment_.Add(new Payment(item));
                            IDManager.SetID(additionalPaymentIDRef_, additionalPayment_[additionalPayment_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            additionalPaymentIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        additionalPayment_.Add(new Payment(item));
                        }
                    }
                    else
                    {
                        additionalPayment_.Add(new Payment(item));
                    }
                }
            }
            
        
            XmlNode earlyTerminationProvisionNode = xmlNode.SelectSingleNode("earlyTerminationProvision");
            
            if (earlyTerminationProvisionNode != null)
            {
                if (earlyTerminationProvisionNode.Attributes["href"] != null || earlyTerminationProvisionNode.Attributes["id"] != null) 
                {
                    if (earlyTerminationProvisionNode.Attributes["id"] != null) 
                    {
                        earlyTerminationProvisionIDRef_ = earlyTerminationProvisionNode.Attributes["id"].Value;
                        EarlyTerminationProvision ob = new EarlyTerminationProvision(earlyTerminationProvisionNode);
                        IDManager.SetID(earlyTerminationProvisionIDRef_, ob);
                    }
                    else if (earlyTerminationProvisionNode.Attributes["href"] != null)
                    {
                        earlyTerminationProvisionIDRef_ = earlyTerminationProvisionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        earlyTerminationProvision_ = new EarlyTerminationProvision(earlyTerminationProvisionNode);
                    }
                }
                else
                {
                    earlyTerminationProvision_ = new EarlyTerminationProvision(earlyTerminationProvisionNode);
                }
            }
            
        
        }
        
    
        #region CapFloorStream_
        private InterestRateStream capFloorStream_;
        public InterestRateStream CapFloorStream_
        {
            get
            {
                if (this.capFloorStream_ != null)
                {
                    return this.capFloorStream_; 
                }
                else if (this.capFloorStreamIDRef_ != null)
                {
                    capFloorStream_ = IDManager.getID(capFloorStreamIDRef_) as InterestRateStream;
                    return this.capFloorStream_; 
                }
                else
                {
                      return this.capFloorStream_; 
                }
            }
            set
            {
                if (this.capFloorStream_ != value)
                {
                    this.capFloorStream_ = value;
                }
            }
        }
        #endregion
        
        public string capFloorStreamIDRef_ { get; set; }
        #region Premium_
        private List<Payment> premium_;
        public List<Payment> Premium_
        {
            get
            {
                if (this.premium_ != null)
                {
                    return this.premium_; 
                }
                else if (this.premiumIDRef_ != null)
                {
                    return this.premium_; 
                }
                else
                {
                      return this.premium_; 
                }
            }
            set
            {
                if (this.premium_ != value)
                {
                    this.premium_ = value;
                }
            }
        }
        #endregion
        
        public string premiumIDRef_ { get; set; }
        #region AdditionalPayment_
        private List<Payment> additionalPayment_;
        public List<Payment> AdditionalPayment_
        {
            get
            {
                if (this.additionalPayment_ != null)
                {
                    return this.additionalPayment_; 
                }
                else if (this.additionalPaymentIDRef_ != null)
                {
                    return this.additionalPayment_; 
                }
                else
                {
                      return this.additionalPayment_; 
                }
            }
            set
            {
                if (this.additionalPayment_ != value)
                {
                    this.additionalPayment_ = value;
                }
            }
        }
        #endregion
        
        public string additionalPaymentIDRef_ { get; set; }
        #region EarlyTerminationProvision_
        private EarlyTerminationProvision earlyTerminationProvision_;
        public EarlyTerminationProvision EarlyTerminationProvision_
        {
            get
            {
                if (this.earlyTerminationProvision_ != null)
                {
                    return this.earlyTerminationProvision_; 
                }
                else if (this.earlyTerminationProvisionIDRef_ != null)
                {
                    earlyTerminationProvision_ = IDManager.getID(earlyTerminationProvisionIDRef_) as EarlyTerminationProvision;
                    return this.earlyTerminationProvision_; 
                }
                else
                {
                      return this.earlyTerminationProvision_; 
                }
            }
            set
            {
                if (this.earlyTerminationProvision_ != value)
                {
                    this.earlyTerminationProvision_ = value;
                }
            }
        }
        #endregion
        
        public string earlyTerminationProvisionIDRef_ { get; set; }
        
    
        
    
    }
    
}

